tool/src/org/antlr/v4/tool/ErrorManager.java

```diff
@@ -216,25 +219,37 @@ public void emit(ErrorType etype, ANTLRMessage msg) {
      *  Otherwise we just use the default "antlr".
      */
     public void setFormat(String formatName) {
-        this.formatName = formatName;
-        String fileName = FORMATS_DIR +formatName+STGroup.GROUP_FILE_EXTENSION;
-        ClassLoader cl = Thread.currentThread().getContextClassLoader();
-        URL url = cl.getResource(fileName);
-        if ( url==null ) {
-            cl = ErrorManager.class.getClassLoader();
-            url = cl.getResource(fileName);
-        }
-        if ( url==null && formatName.equals("antlr") ) {
-            rawError("ANTLR installation corrupted; cannot find ANTLR messages format file "+fileName);
-            panic();
-        }
-        else if ( url==null ) {
-            rawError("no such message format file "+fileName+" retrying with default ANTLR format");
-            setFormat("antlr"); // recurse on this rule, trying the default message format
-            return;
-        }
-        format = new STGroupFile(url, "UTF-8", '<', '>');
-        format.load();
+		STGroupFile loadedFormat = loadedFormats.get(formatName);
+		if (loadedFormat == null) {
```

Comment:
Maybe the situations aren't races but it will still try to load multiple times technically if multiple threads reach the IF at the same time. Is this worth fixing? Since it's not critical code we might want to simply synchronize the method or synchronize on `loadedFormats`.

Reply:
Multiple loading is not a big problem, and it doesn't occur in the code because lock (`synchonized`) is written in the code below. But `HashMap` should be replaced to `ConcurrentHashMap` if it's used in multi-threaded code.

Reply:
> Since it's not critical code we might want to simply synchronize the method or synchronize on loadedFormats.

Reply:
Double-checking locking makes sure that the protected section is being calculated once and allows getting rid of excess locks.

Reply:
The pattern does look familiar...yep. thanks. I missed the double check inside.. thanks.